src/algebra/non_associative/basic.lean

```diff
@@ -0,0 +1,155 @@
+
+end na_algebra
+
+variables [na_semiring A] [na_algebra R A]
+variables [na_semiring B] [na_algebra R B]
+variables [na_semiring C] [na_algebra R C]
+
+set_option old_structure_cmd true
+
+/-- A morphism of non-associative algebras. -/
+structure na_algebra_hom extends A →ₗ[R] B, mul_hom A B
+
+attribute [nolint doc_blame] na_algebra_hom.to_linear_map
+attribute [nolint doc_blame] na_algebra_hom.to_mul_hom
+
+namespace na_algebra_hom
+
+variables {R A B C}
+
+instance : has_coe (na_algebra_hom R A B) (A →ₗ[R] B) := ⟨to_linear_map⟩
+
+instance : has_coe (na_algebra_hom R A B) (mul_hom A B) := ⟨to_mul_hom⟩
+
+/-- see Note [function coercion] -/
+instance : has_coe_to_fun (na_algebra_hom R A B) := ⟨_, to_fun⟩
+
+initialize_simps_projections na_algebra_hom (to_fun → apply)
+
+@[simp] lemma coe_mk (f : A → B) (h₁ h₂ h₃) :
+  ((⟨f, h₁, h₂, h₃⟩ : na_algebra_hom R A B) : A → B) = f := rfl
+
+@[simp, norm_cast] lemma coe_to_linear_map (f : na_algebra_hom R A B) :
+  ((f : A →ₗ[R] B) : A → B) = f :=
+rfl
+
+@[simp, norm_cast] lemma coe_to_mul_hom (f : na_algebra_hom R A B) :
+  ((f : mul_hom A B) : A → B) = f :=
+rfl
+
+@[simp] lemma map_smul (f : na_algebra_hom R A B) (c : R) (x : A) : f (c • x) = c • f x :=
+linear_map.map_smul (f : A →ₗ[R] B) c x
+
+@[simp] lemma map_add (f : na_algebra_hom R A B) (x y : A) : f (x + y) = (f x) + (f y) :=
+linear_map.map_add (f : A →ₗ[R] B) x y
+
+@[simp] lemma map_mul (f : na_algebra_hom R A B) (x y : A) : f (x * y) = (f x) * (f y) :=
+mul_hom.map_mul (f : mul_hom A B) x y
+
+@[simp] lemma map_zero (f : na_algebra_hom R A B) : f 0 = 0 := (f : A →ₗ[R] B).map_zero
+
+instance : has_zero (na_algebra_hom R A B) := ⟨{ map_mul' := by simp, ..(0 : A →ₗ[R] B)}⟩
+
+instance : inhabited (na_algebra_hom R A B) := ⟨0⟩
+
+lemma coe_injective : function.injective (λ f : na_algebra_hom R A B, show A → B, from f) :=
+by rintro ⟨f, _⟩ ⟨g, _⟩ ⟨h⟩; congr
+
+@[ext] lemma ext {f g : na_algebra_hom R A B} (h : ∀ x, f x = g x) : f = g :=
+coe_injective $ funext h
+
+lemma ext_iff {f g : na_algebra_hom R A B} : f = g ↔ ∀ x, f x = g x :=
+⟨by { rintro rfl x, refl }, ext⟩
+
+/-- The composition of morphisms is a morphism. -/
+def comp (f : na_algebra_hom R B C) (g : na_algebra_hom R A B) : na_algebra_hom R A C :=
+{ ..mul_hom.comp (f : mul_hom B C) (g : mul_hom A B),
+  ..linear_map.comp (f : B →ₗ[R] C) (g : A →ₗ[R] B) }
+
+@[simp] lemma comp_apply (f : na_algebra_hom R B C) (g : na_algebra_hom R A B) (x : A) :
+  f.comp g x = f (g x) := rfl
+
+@[norm_cast]
+lemma comp_coe (f : na_algebra_hom R B C) (g : na_algebra_hom R A B) :
+  (f : B → C) ∘ (g : A → B) = f.comp g := rfl
+
+/-- The inverse of a bijective morphism is a morphism. -/
+def inverse (f : na_algebra_hom R A B) (g : B → A)
+  (h₁ : function.left_inverse g f) (h₂ : function.right_inverse g f) : na_algebra_hom R B A :=
+{ map_mul' := λ x y,
+  calc g (x * y) = g ((f (g x)) * (f (g y))) : by { conv_lhs { rw [←h₂ x, ←h₂ y], }, }
```

Comment:
I guess this might be better factored out into a result about `mul_hom`s. Since I'm not yet sure if this PR will be accepted, I thought I'd leave things like this for now to keep it simple.